### ISSUE TITLE
mismatched output from interactive.diffFilter on fresh install

### ISSUE THREAD

**Issue by lizkaraffa** [so-fancy/diff-so-fancy#305]
I've been using diff-so-fancy for years. I messed up some of my node stuff on my mac today and had to do a fresh install of a bunch of npm stuff because gulp stopped working globally. Now that I've updated everything (including npm), my highlighting has gone away completely.  I read about issue #[296](https://github.com/so-fancy/diff-so-fancy/issues/296) but I'm not trying to do anything with patch mode at all. I reinstalled diff-so-fancy with `npm install -g diff-so-fancy` but anytime I try to add my git files with `git add -p file.name` I get this error
```
fatal: mismatched output from interactive.diffFilter
hint: Your filter must maintain a one-to-one correspondence
hint: between its input and output lines.
```
My `.gitconfig` has this in it:
```
[core]
	pager = diff-so-fancy | less --tabs=4 -RFX
[color]
	ui = true
	branch = always
	diff = always
	interactive = always
	status = always
[color "diff"]
	meta = yellow bold
	; whitespace = red reverse
[color "diff-highlight"]
	oldNormal = red bold
	oldHighlight = red bold 52
	newNormal = green bold
	newHighlight = green bold 22

[interactive]
	diffFilter = diff-so-fancy | less -r
```
I'm at a complete loss as to what is going on and how to fix. I cannot imagine working without diff-so-fancy.

I'm running npm version 6.4.0

Thank you!

**Comment by scottchiefbaker** [comment]
I'm sorry I somehow missed this issue when you initially reported it. Were  you able to solve this? I've never used d-s-f in interactive mode, it wouldn't surprise me it doesn't work 100%.

**Comment by lizkaraffa** [comment]
@scottchiefbaker I still have the issue. I used [this](https://github.com/so-fancy/diff-so-fancy/issues/35#issuecomment-394080712) to find a workaround but it's definitely not the awesomeness that used to be when I was using diff-so-fancy. It looks like updating [ubuntu](https://github.com/so-fancy/diff-so-fancy/issues/35#issuecomment-395399301) might be part of the problem, which I'm sure is what happened when I had to fix my machine, I updated everything.

**Comment by scottchiefbaker** [comment]
Going back through old bugs and I see I left you hanging. Were you able to get this fixed? My recommendation would be to try using the script raw from the repo instead of from NPM and see if that solves your problem. If that does we can determine the next best course of action.

**Comment by lizkaraffa** [comment]
@scottchiefbaker unfortunately no. What I said in Feb last year still holds true. Thanks for checking in.

**Comment by scottchiefbaker** [comment]
Does the version from this repo work any differently than the on in NPM?

**Comment by lizkaraffa** [comment]
@scottchiefbaker Thanks again for reaching out. Honestly it's been so long since I messed with this that I'm blanking on what my old workflow used to be. I am not that great on workflow setups and one of my coworkers originally got me going on diff-so-fancy. 

I'm seeing the installation instructions: `Installation is as simple as cloning this repo and then putting the diff-so-fancy script in to your $PATH. The lib/ directory will need to be kept relative to the core script.` but I'm a little unclear where I put the script. Sorry, I have enough knowledge to get myself in trouble.

**Comment by scottchiefbaker** [comment]
If you do `echo $PATH` it will list all the directories that your computer looks in when it tries to run a command.

For example my dev station is:

```
/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/bakers/bin
```

It will check each of those directories in order looking for a command you type. I put d-s-f in `/home/bakers/bin/`. Now whenever I type `diff-so-fancy` (without the path) it loads it from the directory. Otherwise you can put it ANYWHERE and just use the full path when you call it.

**Comment by lizkaraffa** [comment]
@scottchiefbaker when I echoed the $PATH I got this
`/Users/lizkaraffa/.rvm/gems/ruby-2.3.1/bin:/Users/lizkaraffa/.rvm/gems/ruby-2.3.1@global/bin:/Users/lizkaraffa/.rvm/rubies/ruby-2.3.1/bin:/Users/lizkaraffa/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/Users/lizkaraffa/.composer/vendor/bin:/Users/lizkaraffa/.rvm/bin`

I couldn’t figure out how to get into my equivalent of `/home/bakers/bin/` but I thought it might be `Users/lizkaraffa` so I tried to clone the repo there but I was denied access `git@github.com: Permission denied (publickey).` even though my passphrase was correct.

After some more digging I found a`diff-so-fancy` in `usr/local/bin` so I ended up downloading the repo and then swapped out the `diff-so-fancy` exec file I had with the one in the repo.
![Screen Shot 2020-04-30 at 12 51 16 PM](https://user-images.githubusercontent.com/5855545/80752993-5ebde780-8ae1-11ea-800b-fbdee86644d8.png)

Then I quit my terminal, restarted my computer. When I tried `git add -p file.name` I don't see the interactive workflow I used to see. That said, I don't get the fatal error anymore. Instead it's the same visual display as if you did `git a file.name`.

If I can recall correctly, when I used to run `git add -p file.name` it would essentially look like I was running `git diff file.name` + the ability to stage hunks. That was a quick way to visually confirm that what I wanted to stage was correct and allow me to stage only the hunks I wanted. Since I opened this issue I’ve just added an extra step to my workflow where I run `g diff file.name` first and then add the file after I’ve confirmed the code is what I want to stage. The bummer is that when I want to stage hunks I have the not very readable vanilla view of my code.

I have not removed my npm version of `diff-so-fancy` so if that could be causing problems, I can get rid of it. I found it in `Users/lizkaraffa/node_modules/diff-so-fancy`

All this to say, you've definitely tried to do your due diligence. If you want to close this out I totally get it. I miss what I used to have, but it's been so long I don't really remember it all that well anyways.

**Comment by scottchiefbaker** [comment]
No way... I'm not giving up on this :)

For your path you have:

`/Users/lizkaraffa/.rvm/gems/ruby-2.3.1/bin:/Users/lizkaraffa/.rvm/gems/ruby-2.3.1@global/bin:/Users/lizkaraffa/.rvm/rubies/ruby-2.3.1/bin:/Users/lizkaraffa/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/Users/lizkaraffa/.composer/vendor/bin:/Users/lizkaraffa/.rvm/bin`

Your system uses each one of those directories, in order when trying to find a command. Ignoring the first couple as they're all specific to ruby I see `/Users/lizkaraffa/bin`. That seems like a good place to put `diff-so-fancy` and the `lib/` directory.

You can run `which diff-so-fancy` and it will tell you what directory it would run that command from. This is useful in cases where you may have a d-s-f in two places. Like say NPM installed one, and you manually install one from GitHub.

**Comment by OJFord** [comment]
This is the sticking point in #35 isn't it? I think since some git version in the 2.teens diff-so-fancy can't be used as an `interactive.diffFilter`, since git needs 1:1 correspondance between the filtered lines, and shuffles them _after_ filtering, so [prettifying them doesn't work](https://github.com/so-fancy/diff-so-fancy/issues/35#issuecomment-229208924) (such as you tried with a `--patch-mode` option in #172).

I came across `delta` recently which solves this with a `--color-only` (or something like that) option, which for dsf is equivalent to just `diff-highlight`, and that's what most ended up using as a result in that thread I believe.

Maybe it'd be worth reviving `--patch-mode` @scottchiefbaker, even if all it does (all it can do) is delegate to `diff-highlight` and not touch the headers, just to avoid all this confusion over whether `add -p` is 'supported', what it means to support it, and to what extent that's even possible. (In the HN thread that made me aware of delta there was some discussion that 'unlike dsf' it supports `add -p`; actually it just does its colouring, since for the same reason that's all it can do, and for dsf that's handled by `diff-highlight`, so there's nothing to do. I did point this out.)

**Comment by wren** [comment]
Is this issue fixed now that #35 is closed?

**Comment by scottchiefbaker** [comment]
I believe this should be fixed since we added `--patch` mode.

**Comment by OJFord** [comment]
There seems to be a `diff-highlight` bug when the diff is (or includes) newline/no newline at the end of the file. So we may actually see more reports of this from people using `--patch`, but it seems to come from `diff-highlight` (i.e. I can reproduce it with just `diff-highlight` as my `interactive.diffFilter`).

**Comment by milianw** [comment]
I just ran into this issue too. None of the hunks are at the end of the file though, so it seems to be something else on top which can break this feature.

**Comment by wren** [comment]
@OJFord @milianw Do you have an example to repro this? I'd be happy to take a look to see if it's resolvable in d-s-f.

**Comment by milianw** [comment]
For me it's as easy as this:

Setup diff-so-fancy according to docs:
```
pacman -S diff-so-fancy
git config --global core.pager "diff-so-fancy | less --tabs=4 -RFX"
git config --global interactive.diffFilter "diff-so-fancy --patch"
```

Then try it out:
```
git clone git@github.com:so-fancy/diff-so-fancy.git
cd diff-so-fancy
git reset HEAD^
git add -p
```

You will be greated by:
```
fatal: mismatched output from interactive.diffFilter
hint: Your filter must maintain a one-to-one correspondence
hint: between its input and output lines.
```

**Comment by OJFord** [comment]
@milianw You need `diff-so-fancy-git` from the AUR for v1.4.0 with `--patch` support; `community/diff-so-fancy` is out of date.

**Comment by milianw** [comment]
Ah, thank you - that fixes it.

**Comment by wren** [comment]
@OJFord Do you have an example of the issue you mentioned above?

**Comment by OJFord** [comment]
@wren I could've sworn it was as simple as add/removing a trailing newline and `git config interactive.diffFilter diff-highlight ; git add -p`, but I can't reproduce it now. 

_Maybe_ this fixed it? https://github.com/git/git/commit/579789dbced8f9aee987fc4d609f5f92d32f9d26 (It is suspiciously about 'blank lines', but the actual motivation behind it is completely unrelated; I'm not familiar with it to know if it might be hitting the same path in this case.)

But that was released in v2.29.0, and v.2.30.x was available when I commented about it above; I'd be surprised if I was a few months out of date, but I suppose it's possible - I was away with my laptop at the time, which I don't keep so up to date.

Anyway, I can no longer reproduce it with `diff-highlight` from `git --version` v2.31.1.

**Comment by OJFord** [comment]
Closing as this is now supported (#390, thanks @wren!) via the `--patch` flag; your git config should contain:
```ini
[interactive]
    diffFilter = diff-so-fancy --patch
```
**in addition to** the `pager.diff = diff-so-fancy | less ...` line for non-patch-mode behaviour.

If anyone sees this 'mismatched output' issue now **using v1.4.0 or later** please open a new issue with diff-so-fancy & git `--version`s, git config, and a diff that reproduces the error.

Thanks!
